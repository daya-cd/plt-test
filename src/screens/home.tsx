import React, {FC, useCallback, useEffect, useState} from 'react';
import {SafeAreaView, StyleSheet, View, Text, FlatList} from 'react-native';
import {Subheading} from 'react-native-paper';
import Products from '../components/products';
import {IProduct} from '../shared/interfaces/product';

const Home: FC = () => {
  const [products, setProducts] = useState<IProduct[]>([]);
  const [error, setError] = useState<string | null>();

  const fetcData = async () => {
    const resp = await fetch(
      'https://my-json-server.typicode.com/benirvingplt/products/products',
    );
    if (!resp.ok) {
      setError('Something went wrong');
    } else {
      const data = await resp.json();
      setProducts(data);
      setError(null);
    }
  };

  useEffect(() => {
    fetcData();
  }, []);

  const renderItem = useCallback(({item}) => <Products product={item} />, []);
  const keyExtractor = useCallback(item => item.id.toString(), []);

  return (
    <SafeAreaView style={styles.container}>
      <View>
        {!products && <Text> Loading...</Text>}
        {products && (
          <FlatList
            data={products}
            keyExtractor={keyExtractor}
            renderItem={renderItem}
          />
        )}
        {error && (
          <View style={styles.noProduct}>
            <Subheading>Something went wrong in API call</Subheading>
          </View>
        )}
      </View>
    </SafeAreaView>
  );
};

const styles = StyleSheet.create({
  container: {
    flex: 1,
  },
  noProduct: {
    margin: 20,
  },
});

export default Home;
