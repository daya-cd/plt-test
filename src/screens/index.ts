import Home from './home';
import Basket from './basket';
import ProductDetail from './productDetail';

export {Home, Basket, ProductDetail};
