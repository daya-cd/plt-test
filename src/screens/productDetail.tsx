import React from 'react';
import {View, SafeAreaView, StyleSheet, Image} from 'react-native';
import {
  Card,
  Title,
  Button,
  Portal,
  Dialog,
  Paragraph,
} from 'react-native-paper';
import {useDispatch} from 'react-redux';
import {bindActionCreators} from 'redux';
import {actionCreators} from '../shared/store';
import {IProduct} from '../shared/interfaces/product';

const ProductDetail = ({route}) => {
  const dispatch = useDispatch();
  const {addProduct} = bindActionCreators(actionCreators, dispatch);
  const [visible, setVisible] = React.useState(false);

  const hideDialog = () => setVisible(false);

  const addToCart = (product: IProduct): void => {
    addProduct(product);
    setVisible(true);
  };
  const {productDetail} = route.params;
  return (
    <SafeAreaView style={styles.container}>
      <View>
        <Image
          source={{uri: productDetail.img}}
          style={styles.imageContainer}
          resizeMode="center"
        />
        <Card.Title
          titleNumberOfLines={2}
          titleStyle={styles.title}
          title={productDetail.name}
          subtitle={'Colour: ' + productDetail.colour}
          right={() => <Title>Â£{productDetail.price}</Title>}
          rightStyle={styles.rightStyle}
        />
        <Button
          mode="contained"
          testID="addCartButton"
          onPress={() => addToCart(productDetail)}>
          ADD TO CART
        </Button>
        <Portal>
          <Dialog visible={visible} onDismiss={hideDialog}>
            <Dialog.Title>Update:</Dialog.Title>
            <Dialog.Content>
              <Paragraph>Item has been added to your basket.</Paragraph>
            </Dialog.Content>
            <Dialog.Actions>
              <Button onPress={hideDialog}>Done</Button>
            </Dialog.Actions>
          </Dialog>
        </Portal>
      </View>
    </SafeAreaView>
  );
};

export default ProductDetail;

const styles = StyleSheet.create({
  container: {
    flex: 1,
    margin: 10,
  },
  imageContainer: {
    height: 400,
  },
  title: {
    fontSize: 14,
  },
  rightStyle: {
    paddingRight: 20,
  },
});
