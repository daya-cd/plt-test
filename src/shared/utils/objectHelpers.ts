const isEmpty = (obj: object) => {
  return Object.keys(obj).length === 0;
};
export {isEmpty};
