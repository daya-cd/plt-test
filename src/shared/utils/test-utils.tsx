import React, {ReactElement} from 'react';
import {
  render as rtlRender,
  RenderOptions,
} from '@testing-library/react-native';
import {Provider} from 'react-redux';
import {Provider as PaperProvider} from 'react-native-paper';
import {store} from '../store';

export interface WrapperProps {
  children: ReactElement;
}

const render = (ui: ReactElement, renderOptions?: RenderOptions) => {
  const Wrapper = ({children}: WrapperProps): ReactElement => {
    return (
      <Provider store={store}>
        <PaperProvider>{children}</PaperProvider>
      </Provider>
    );
  };
  return rtlRender(ui, {wrapper: Wrapper, ...renderOptions});
};

// re-export everything
export * from '@testing-library/react-native';

// override render method
export {render};
