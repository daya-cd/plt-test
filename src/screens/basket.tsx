/* eslint-disable no-shadow */
import React from 'react';
import {
  SafeAreaView,
  StyleSheet,
  Image,
  ScrollView,
  TouchableOpacity,
  View,
} from 'react-native';
import {Colors, List, Subheading} from 'react-native-paper';
import {useSelector, useDispatch} from 'react-redux';
import {RootState} from '../shared/store/reducers';
import {bindActionCreators} from 'redux';
import {actionCreators} from '../shared/store';
import {IProduct} from '../shared/interfaces/product';
import {isEmpty} from '../shared/utils/objectHelpers';

const Basket = () => {
  const state = useSelector((state: RootState) => state.shop);
  const dispatch = useDispatch();
  const {removeProduct} = bindActionCreators(actionCreators, dispatch);
  const removeItemFromCart = (product: IProduct): void => {
    removeProduct(product);
  };
  if (isEmpty(state.items)) {
    return (
      <View style={styles.noProduct}>
        <Subheading>
          No products in the basket. Please select from home
        </Subheading>
      </View>
    );
  }
  return (
    <SafeAreaView style={styles.container}>
      <ScrollView>
        {state &&
          state.items.map((items, i) => (
            <List.Item
              key={i}
              title={items.name}
              titleNumberOfLines={2}
              description={
                'Price: Â£' + items.price + '  ' + 'Colour: ' + items.colour
              }
              left={() => (
                <Image
                  source={{uri: items.img}}
                  style={styles.imageContainer}
                  resizeMode="contain"
                />
              )}
              right={props => (
                <TouchableOpacity
                  onPress={() => removeItemFromCart(items)}
                  testID="deleteIcon">
                  <List.Icon {...props} color={Colors.redA700} icon="delete" />
                </TouchableOpacity>
              )}
            />
          ))}
      </ScrollView>
    </SafeAreaView>
  );
};

const styles = StyleSheet.create({
  container: {
    flex: 1,
  },
  imageContainer: {
    width: 80,
    height: 80,
  },
  noProduct: {
    margin: 20,
  },
});

export default Basket;
