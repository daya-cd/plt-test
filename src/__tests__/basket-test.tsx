/**
 * @format
 */

import 'react-native';
import React from 'react';
import * as redux from 'react-redux';
import {Basket} from '../screens';
import {cleanup, fireEvent, render} from '../shared/utils/test-utils';

describe('renders basket view', () => {
  const useSelectorMock = jest.spyOn(redux, 'useSelector');
  const useDispatchMock = jest.spyOn(redux, 'useDispatch');
  beforeEach(() => {
    useSelectorMock.mockClear();
    useDispatchMock.mockClear();
  });
  afterEach(cleanup);
  useSelectorMock.mockReturnValue({
    items: [
      {
        id: 1,
        colour: 'Black',
        name: 'Black Sheet Strappy Textured Glitter Bodycon Dress',
        price: 10,
        img: 'http://cdn-img.prettylittlething.com/9/0/a/a/90aa90903a135ee59594f47c7685aa7ef3046e44_cly8063_1.jpg?imwidth=1024',
      },
    ],
  });
  it('renders correctly', () => {
    const tree = render(<Basket />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders product name correctly', async () => {
    const {findByText} = render(<Basket />);
    const productName = await findByText(
      /Black Sheet Strappy Textured Glitter Bodycon Dress/i,
    );
    expect(productName).toBeTruthy();
  });
  it('renders price information correctly', async () => {
    const {findByText} = render(<Basket />);
    const productPrice = await findByText(/10/i);
    expect(productPrice).toBeTruthy();
  });
  it('fires removeProduct dispatch action', () => {
    const removeProduct = jest.fn();
    useDispatchMock.mockReturnValue(removeProduct);
    const {getByTestId} = render(<Basket />);
    expect(removeProduct).not.toHaveBeenCalled();
    fireEvent.press(getByTestId('deleteIcon'));
    expect(removeProduct).toHaveBeenCalled();
  });
  it('renders no product text if basket is empty', async () => {
    useSelectorMock.mockReturnValue({items: []});
    const {findByText} = render(<Basket />);
    const productPrice = await findByText(
      /No products in the basket. Please select from home/i,
    );
    expect(productPrice).toBeTruthy();
  });
});
