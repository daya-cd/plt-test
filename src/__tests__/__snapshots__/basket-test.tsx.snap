// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders basket view renders correctly 1`] = `
<View
  collapsable={false}
  pointerEvents="box-none"
  style={
    Object {
      "flex": 1,
    }
  }
>
  <RCTSafeAreaView
    emulateUnlessSupported={true}
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <RCTScrollView>
      <View>
        <View
          accessibilityState={
            Object {
              "disabled": true,
            }
          }
          accessible={true}
          focusable={false}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Array [
              false,
              Array [
                Object {
                  "padding": 8,
                },
                undefined,
              ],
            ]
          }
        >
          <View
            style={
              Object {
                "flexDirection": "row",
              }
            }
          >
            <Image
              resizeMode="contain"
              source={
                Object {
                  "uri": "http://cdn-img.prettylittlething.com/9/0/a/a/90aa90903a135ee59594f47c7685aa7ef3046e44_cly8063_1.jpg?imwidth=1024",
                }
              }
              style={
                Object {
                  "height": 80,
                  "width": 80,
                }
              }
            />
            <View
              style={
                Array [
                  Object {
                    "marginVertical": 6,
                    "paddingLeft": 8,
                  },
                  Object {
                    "flex": 1,
                    "justifyContent": "center",
                  },
                ]
              }
            >
              <Text
                numberOfLines={2}
                selectable={false}
                style={
                  Array [
                    Object {
                      "color": "#000000",
                      "fontFamily": "System",
                      "fontWeight": "400",
                    },
                    Object {
                      "textAlign": "left",
                    },
                    Array [
                      Object {
                        "fontSize": 16,
                      },
                      Object {
                        "color": "rgba(0, 0, 0, 0.87)",
                      },
                      undefined,
                    ],
                  ]
                }
              >
                Black Sheet Strappy Textured Glitter Bodycon Dress
              </Text>
              <Text
                numberOfLines={2}
                selectable={false}
                style={
                  Array [
                    Object {
                      "color": "#000000",
                      "fontFamily": "System",
                      "fontWeight": "400",
                    },
                    Object {
                      "textAlign": "left",
                    },
                    Array [
                      Object {
                        "fontSize": 14,
                      },
                      Object {
                        "color": "rgba(0, 0, 0, 0.54)",
                      },
                      undefined,
                    ],
                  ]
                }
              >
                Price: £10  Colour: Black
              </Text>
            </View>
            <View
              accessible={true}
              collapsable={false}
              focusable={true}
              nativeID="animatedComponent"
              onClick={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                Object {
                  "opacity": 1,
                }
              }
              testID="deleteIcon"
            >
              <View
                pointerEvents="box-none"
                style={
                  Array [
                    Object {
                      "alignItems": "center",
                      "height": 40,
                      "justifyContent": "center",
                      "margin": 8,
                      "width": 40,
                    },
                    Object {
                      "marginRight": 0,
                    },
                  ]
                }
              >
                <Text
                  accessibilityElementsHidden={true}
                  allowFontScaling={false}
                  importantForAccessibility="no-hide-descendants"
                  pointerEvents="none"
                  selectable={false}
                  style={
                    Array [
                      Object {
                        "color": "#d50000",
                        "fontSize": 24,
                      },
                      Array [
                        Object {
                          "lineHeight": 24,
                          "transform": Array [
                            Object {
                              "scaleX": 1,
                            },
                          ],
                        },
                        Object {
                          "backgroundColor": "transparent",
                        },
                      ],
                      Object {
                        "fontFamily": "Material Design Icons",
                        "fontStyle": "normal",
                        "fontWeight": "normal",
                      },
                      Object {},
                    ]
                  }
                >
                  󰆴
                </Text>
              </View>
            </View>
          </View>
        </View>
      </View>
    </RCTScrollView>
  </RCTSafeAreaView>
</View>
`;
